cookbook/slackbot/start.py

```diff
@@ -1,17 +1,82 @@
-from handler import handle_message
-from marvin import AIApplication
-from marvin.deployment import Deployment
-
-deployment = Deployment(
-    component=AIApplication(tools=[handle_message]),
-    app_kwargs={
-        "title": "Marvin Slackbot",
-        "description": "A Slackbot powered by Marvin",
-    },
-    uvicorn_kwargs={
-        "port": 4200,
-    },
-)
+import asyncio
+import re
+
+import uvicorn
+from cachetools import TTLCache
+from fastapi import FastAPI, HTTPException, Request
+from marvin import Assistant
+from marvin.beta.assistants import Thread
+from marvin.tools.github import search_github_issues
+from marvin.tools.retrieval import multi_query_chroma
+from marvin.utilities.logging import get_logger
+from marvin.utilities.slack import SlackPayload, post_slack_message
+from prefect import flow, task
+from prefect.states import Completed
+
+app = FastAPI()
+BOT_MENTION = r"<@(\w+)>"
+CACHE = TTLCache(maxsize=100, ttl=86400 * 7)
+
+
+@flow
+async def handle_message(payload: SlackPayload):
+    logger = get_logger("slackbot")
+    user_message = (event := payload.event).text
+    cleaned_message = re.sub(BOT_MENTION, "", user_message).strip()
+    logger.debug_kv("Handling slack message:", user_message, "green")
+    if (user := re.search(BOT_MENTION, user_message)) and user.group(
+        1
+    ) == payload.authorizations[0].user_id:
+        thread = event.thread_ts or event.ts
+        assistant_thread = CACHE.get(thread, Thread())
+        CACHE[thread] = assistant_thread
+
+        with Assistant(
+            name="Marvin (from Hitchhiker's Guide to the Galaxy)",
+            tools=[task(multi_query_chroma), task(search_github_issues)],
+            instructions=(
+                "use chroma to search docs and github to search"
+                " issues and answer questions about prefect 2.x."
+                " you must use your tools in all cases except where"
+                " the user simply wants to converse with you."
+            ),
+        ) as assistant:
+            await assistant_thread.add_async(cleaned_message)
+            run = await assistant_thread.run_async(assistant)
+            await task(post_slack_message)(
+                ai_response_text := run.thread.get_messages()[-1].content[0].text.value,
```

Comment:
Interesting quirk about the assistants API -- assistants can post multiple messages to the thread in a single "run". Here's one approach you might take to getting them all. Note if for some reason one of them was an image response the `content[0].text.value` could fail but for this bot I dont think thats an issue

```python
# get the reference to the user message object
user_message = await assistant_thread.add_async(cleaned_message)
run = await assistant_thread.run_async(assistant)
# get all messages that were posted after the user message
ai_messages = assistant_thread.get_messages(after_message=user_message.id)
# extract content
ai_messages_content = [m.content[0].text.value for m in ai_messages]
await task(post_slack_message)(
    ai_response_text '\n\n'.join(ai_messages_content),
    ...
```

Reply:
cool, thanks! added in [42bca9c](https://github.com/PrefectHQ/marvin/pull/658/commits/42bca9c49221c92c851cc874390a96fb8420c276)